hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/request/validation/ValidatorRegistry.java

```diff
@@ -0,0 +1,297 @@
+                  "Allowed versionType: %4$s", valueName, methodName, value.toString(), values));
+    }
+  }
+
+  /**
+   * Initializes the internal request validator store.
+   * The requests are stored in the following structure:
+   * - An EnumMap with the RequestType as the key, and in which
+   *   - values are an EnumMap with the request processing phase as the key, and in which
+   *     - values is an {@link IndexedItems } containing the validation list
+   * @param validatorsToBeRegistered collection of the annotated validtors to process.
+   */
+  private void initMaps(Class<RequestType[]> requestType,
+                        Set<RequestProcessingPhase> allowedPhases,
+                        Collection<Class<? extends Annotation>> validatorsToBeRegistered,
+                        Reflections reflections) {
+    for (Class<? extends Annotation> validator : validatorsToBeRegistered) {
+      registerValidator(requestType, allowedPhases, validator, reflections);
+    }
+  }
+
+  private void registerValidator(Class<RequestType[]> requestType,
+                                 Set<RequestProcessingPhase> allowedPhases,
+                                 Class<? extends Annotation> validatorToBeRegistered,
+                                 Reflections reflections) {
+    Collection<Method> methods =  reflections.getMethodsAnnotatedWith(validatorToBeRegistered);
+    Class<? extends Version> versionClass = (Class<? extends Version>)
+        this.getReturnTypeOfAnnotationMethod(validatorToBeRegistered, RegisterValidator.APPLY_UNTIL_METHOD_NAME);
+    List<Pair<? extends Annotation, Method>> sortedMethodsByApplyUntilVersion = methods.stream()
+        .map(method -> Pair.of(method.getAnnotation(validatorToBeRegistered), method))
+        .sorted((validatorMethodPair1, validatorMethodPair2) ->
+            Integer.compare(
+                this.getApplyUntilVersion(validatorMethodPair1.getKey()).version(),
+                this.getApplyUntilVersion(validatorMethodPair2.getKey()).version()))
+        .collect(Collectors.toList());
+    for (Pair<? extends Annotation, Method> validatorMethodPair : sortedMethodsByApplyUntilVersion) {
+      Annotation validator = validatorMethodPair.getKey();
+      Method method = validatorMethodPair.getValue();
+      Version applyUntilVersion = this.getApplyUntilVersion(validator);
+      RequestProcessingPhase phase = this.getRequestPhase(validator);
+      checkAllowedAnnotationValues(allowedPhases, phase, RegisterValidator.PROCESSING_PHASE_METHOD_NAME,
+          method.getName());
+      Set<RequestType> types = Sets.newHashSet(this.getRequestType(validator, requestType));
+      method.setAccessible(true);
+      for (RequestType type : types) {
+        EnumMap<RequestType, EnumMap<RequestProcessingPhase, IndexedItems<Method, Integer>>> requestMap =
+            this.indexedValidatorMap.get(versionClass);
+        EnumMap<RequestProcessingPhase, IndexedItems<Method, Integer>> phaseMap =
```

Comment:
ApplyUntilVersion is of integer value type. This represents 2 type version enum version:
1. Client Version
2. Layout Version
Using integer, it can not represent both and there is chance of picking wrong method. So we need have either separate storage for each type or some way to distinguish them while filtering during validation.

cc: @fapifta

Reply:
The indexedValidatorMap has different mappings for different version classes at the top level, that makes the distinction possible.

The background of this is more interesting though...
We have a proto definition that defines messages between a server and a client. In this case the server is the OM, while the client is the Ozone client. Both of these parties have a version associated with them, in this case the ClientVersion, and the OmLayoutVersion.

Both sides can be aware of the corresponding version of the other side within its release version (though afaik the Ozone client does not care or may not know about the OmLayoutVersion as we did not handle compatibility from that side).

Now...
When an older client sends a request to a newer server, then the server has its client version hardcoded in its codebase, and compares that with the client version in the request. If the client version is different, it selects the validators that are required for the client version supplied in the request from the client version type related mappings.
On top of this, the server knows is hardcoded server version, and its finalized server version, and if there is a difference between the two, then we are in a pre-finalized state, and it selects the validators that has to be applied for the finalized server version it has.

When a newer client sends a request, the client version translates to FUTURE_VERSION, as the older server does not have the new client version enum value. In this case it does not apply any validators on the client version, but it still applies server version validators if there are any non-finalized server versions. (Like in a case where server has a version released in Ozone 3.2, the client is using code released with version 3.3, while the server side was just updated from 3.1 to 3.2.)


Why we have validations for both the server and the client version?
There may be changes in the metadata layout on the server side that does not affect the client side, so the client version is not bumped, but the server version is bumped. In this case we need the server side version to properly finalize things.
But on the other hand after the finalization, we need to validate the client version, as older clients may not understand data returned in the response, as that may contain newly added things. In this case we either need to send an exception to the client that informs the client about the inability to serve the response to an old client, or we need to translate the data so that the old client understands. This part we can not spare, and after finalizing an upgrade old clients are affected.


So that is why we have a different set of validators based on versions, as one case is to handle older clients, the other is to handle the pre finalized state.


In the OM and Ozone client relation for which we have the system set up, OMClientVersionValidator is dealing with old clients, while the OMLayoutFeatureValidator deals with the pre-finalized state of OM.